### PR TITLE
README.md + position_velocity_effort_gain_controller + copy of system_bolt_interface.hpp in gazebo_ros_control_bolt

### PR DESCRIPTION
**Pull request by Maxime-Fansi-laas**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,39 +1,26 @@
...
+# ROS2_CONTROL_BOLT
 
-[![Build Status](https://github.com/ros-controls/ros2_control_demos/workflows/CI/badge.svg?branch=master)](https://github.com/ros-controls/ros2_control_demos/actions?query=workflow%3ACI)
-[![Linters Status](https://github.com/ros-controls/ros2_control_demos/workflows/Linters/badge.svg?branch=master)](https://github.com/ros-controls/ros2_control_demos/actions?query=workflow%3ALinters)
-[![Coverage Status](https://github.com/ros-controls/ros2_control_demos/workflows/Coverage/badge.svg?branch=master)](https://github.com/ros-controls/ros2_control_demos/actions?query=workflow%3ACoverage)
-[![Licence](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
+## Introduction
 
-This repository provides templates for the development of `ros2_control`-enabled robots and a simple simulations to demonstrate and prove `ros2_control` concepts.
+`ros2_control_bolt` is a repository that enables designing and simulation of Bolt robot for testing its controller.
```
Comment (olivier-stasse) [inline README.md]: designing and simulation of Bolt -> controlling and simulating the Bolt robot

File: README.md [hunk 2]
```diff
@@ -1,39 +1,26 @@
...
-2. The second package, `ros2_control_demo_description`, stored URDF-description files, rviz configurations and meshes for the demo robots.
-3. The third package, `ros2_control_demo_hardware`, implements the hardware interfaces described in the roadmap.
-The examples simulate a simple *RRbot* internally to provide sufficient test and demonstration data and reduce external dependencies.
-This package does not have any dependencies except `ros2` core packages and can, therefore, be used on SoC-hardware of headless systems.
+`position_velocity_effort_gain_controller` - a ROS2 package in which is implemented a simple controller that enables to send position, velocity, effort and gain commands to the robot, with the specific hardware interface.
 
-This repository demonstrates the following `ros2_control` concepts:
+`ros2_control_bolt_bringup` - contains files that enable to launch Bolt within its GUIs, and with it hardware interfaces and controllers.
+
+`ros2_control_test_nodes` - contains nodes for testing ros2_control-enabled robots and their integration in the framework.
```
Comment (olivier-stasse) [inline README.md]: Maybe this package should be removed if this is the same than the initial `ros2_control_demos`repository.

File: README.md [hunk 3]
```diff
@@ -51,274 +38,32 @@ These are some quick hints, especially for those coming from a ROS1 control back
   The ros2_control framework uses the **<ros2_control>** tag in the URDF.
 * Joint names in <ros2_control> tags in the URDF must be compatible with the controller's configuration.
```
Comment (olivier-stasse) [inline README.md]: Those two lines do not add anything very informative for our repository.

File: gazebo_ros2_control_bolt/include/gazebo_ros2_control_bolt/gazebo_bolt_system_interface.hpp [hunk 4]
```diff
@@ -30,6 +30,8 @@
 
 #include "rclcpp/rclcpp.hpp"
 
+// #include <system_interface_bolt.hpp>
```
Comment (olivier-stasse) [inline gazebo_ros2_control_bolt/include/gazebo_ros2_control_bolt/gazebo_bolt_system_interface.hpp]: Please remove the comments to make the code leaner.

File: gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp [hunk 5]
```diff
@@ -428,38 +472,36 @@ hardware_interface::return_type GazeboBoltSystem::read()
...
      RCLCPP_INFO_STREAM(this->nh_->get_logger(), "\tjoint: "  << this->dataPtr->sim_joints_[j]->GetName());
 
     if (this->dataPtr->sim_joints_[j]) {
       if (this->dataPtr->joint_control_methods_[j] & POSITION) {
-        RCLCPP_INFO_STREAM(this->nh_->get_logger(), "");   //"\tis controlled in POSITION"
-        /*this->dataPtr->sim_joints_[j]->SetPosition(
-          0, this->dataPtr->joint_position_cmd_[j],
-          true);*/
-          RCLCPP_INFO_STREAM(this->nh_->get_logger(),dataPtr->joint_position_cmd_[j] );
+        RCLCPP_INFO_STREAM(this->nh_->get_logger(), "position");   //"\tis controlled in POSITION"
```
Comment (olivier-stasse) [inline gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp]: Please change it to RCLCPP_DEBUG_STREAM

File: gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp [hunk 6]
```diff
@@ -428,38 +472,36 @@ hardware_interface::return_type GazeboBoltSystem::read()
...
+        /*this->dataPtr->sim_joints_[j]->SetPosition(0, this->dataPtr->joint_position_cmd_[j],true);*/
+          RCLCPP_INFO_STREAM(this->nh_->get_logger(),"");
       }
       if (this->dataPtr->joint_control_methods_[j] & VELOCITY) {
-        RCLCPP_INFO_STREAM(this->nh_->get_logger(), "");    //\tis controlled in VELOCITY
-        /*this->dataPtr->sim_joints_[j]->SetVelocity(
-          0,
-          this->dataPtr->joint_velocity_cmd_[j]);*/
-         RCLCPP_INFO_STREAM(this->nh_->get_logger(),dataPtr->joint_velocity_cmd_[j] );
+        RCLCPP_INFO_STREAM(this->nh_->get_logger(), "velocity");    //\tis controlled in VELOCITY
```
Comment (olivier-stasse) [inline gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp]: Please change it to RCLCPP_DEBUG_STREAM

File: gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp [hunk 7]
```diff
@@ -428,38 +472,36 @@ hardware_interface::return_type GazeboBoltSystem::read()
...
+        /*this->dataPtr->sim_joints_[j]->SetVelocity(0,this->dataPtr->joint_velocity_cmd_[j]);*/
+        // this->dataPtr->joint_velocity_[j] =
+        //   this->dataPtr->joint_kd_cmd_[j]*(this->dataPtr->joint_velocity_cmd_[j] - this->dataPtr->joint_velocity_[j]);
+        // this->dataPtr->sim_joints_[j]->SetVelocity(0, this->dataPtr->joint_velocity_[j]);
```
Comment (olivier-stasse) [inline gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp]: Please remove the comments to make the code leaner.

File: gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp [hunk 8]
```diff
@@ -428,38 +472,36 @@ hardware_interface::return_type GazeboBoltSystem::read()
...
+         RCLCPP_INFO_STREAM(this->nh_->get_logger(),  "");
```
Comment (olivier-stasse) [inline gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp]: Please change it to RCLCPP_DEBUG_STREAM

File: gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp [hunk 9]
```diff
@@ -428,38 +472,36 @@ hardware_interface::return_type GazeboBoltSystem::read()
...
       }
       if (this->dataPtr->joint_control_methods_[j] & EFFORT) {
-        RCLCPP_INFO_STREAM(this->nh_->get_logger(), "");   //"\tis controlled in EFFORT"
-        const double effort =
-          kp*(this->dataPtr->joint_position_cmd_[j] - this->dataPtr->joint_position_[j]) +
-          kd*(this->dataPtr->joint_velocity_cmd_[j] - this->dataPtr->joint_velocity_[j]);
-        this->dataPtr->sim_joints_[j]->SetForce(0, effort);
-        RCLCPP_INFO_STREAM(this->nh_->get_logger(), effort );
+        RCLCPP_INFO_STREAM(this->nh_->get_logger(), "effort");   //"\tis controlled in EFFORT"
```
Comment (olivier-stasse) [inline gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp]: Please change it to RCLCPP_DEBUG_STREAM

File: gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp [hunk 10]
```diff
@@ -428,38 +472,36 @@ hardware_interface::return_type GazeboBoltSystem::read()
...
+        this->dataPtr->joint_effort_[j] = this->dataPtr->joint_effort_cmd_[j] +
+          this->dataPtr->joint_kp_cmd_[j]*(this->dataPtr->joint_position_cmd_[j] - this->dataPtr->joint_position_[j]) +
+          this->dataPtr->joint_kd_cmd_[j]*(this->dataPtr->joint_velocity_cmd_[j] - this->dataPtr->joint_velocity_[j]);
+        this->dataPtr->sim_joints_[j]->SetForce(0,  this->dataPtr->joint_effort_[j]);
+        RCLCPP_INFO_STREAM(this->nh_->get_logger(), this->dataPtr->joint_effort_[j]);
```
Comment (olivier-stasse) [inline gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp]: Please change it to RCLCPP_DEBUG_STREAM

File: position_velocity_effort_gain_controller/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  find_package(std_msgs REQUIRED)
+
+
+#2. compile directive for a shared library providing the cpp file as the source
+add_library(${PROJECT_NAME}
+  SHARED
+  src/position_velocity_effort_gain_controller.cpp
+)
+
+#3.
```
Comment (olivier-stasse) [inline position_velocity_effort_gain_controller/CMakeLists.txt]: Adding comments is a very good idea.

The line below is adding the repository's `include` directory in the set of directories where headers are looking for. This happens in the code with `#include` cpp macro.

File: position_velocity_effort_gain_controller/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+  realtime_tools
+  std_msgs
+)
+
+#5. pluginlib description
+target_compile_definitions(${PROJECT_NAME} PRIVATE "POSITION_VELOCITY_EFFORT_GAIN_CONTROLLER_BUILDING_DLL")
+pluginlib_export_plugin_description_file(${PROJECT_NAME} position_velocity_effort_gain_controller_plugins.xml)
+# prevent pluginlib from using boost
+target_compile_definitions(${PROJECT_NAME} PUBLIC "PLUGINLIB__DISABLE_BOOST_FUNCTIONS")
+pluginlib_export_plugin_description_file(controller_interface position_velocity_effort_gain_controller_plugins.xml)
```
Comment (olivier-stasse) [inline position_velocity_effort_gain_controller/CMakeLists.txt:52]: Could you explain why you need to have two exports for the same XML file ?

File: position_velocity_effort_gain_controller/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    test_position_velocity_effort_gain_controller
+    test/test_position_velocity_effort_gain_controller.cpp
+  )
+  target_include_directories(test_position_velocity_effort_gain_controller PRIVATE include)
+  target_link_libraries(test_position_velocity_effort_gain_controller
+    ${PROJECT_NAME}
+  )
+
+
+endif()
```
Comment (olivier-stasse) [inline position_velocity_effort_gain_controller/CMakeLists.txt:96]: Could you add the option to disable `BUILD_TESTING` by default ?
That should be:
`option(BUILD_TESTING "Building tests" OFF)`

File: position_velocity_effort_gain_controller/test/test_load_position_velocity_effort_gain_controller.cpp [hunk 14]
```diff
@@ -0,0 +1,39 @@
+// Copyright 2020 PAL Robotics SL.
```
Comment (olivier-stasse) [inline position_velocity_effort_gain_controller/test/test_load_position_velocity_effort_gain_controller.cpp]: Please add
`// Copyright 2022 LAAS CNRS`

File: position_velocity_effort_gain_controller/test/test_position_velocity_effort_gain_controller.cpp [hunk 15]
```diff
@@ -0,0 +1,232 @@
+// Copyright 2020 PAL Robotics SL.
```
Comment (olivier-stasse) [inline position_velocity_effort_gain_controller/test/test_position_velocity_effort_gain_controller.cpp]: Please add the line:
`// Copyright 2022 LAAS CNRS`

File: position_velocity_effort_gain_controller/test/test_position_velocity_effort_gain_controller.hpp [hunk 16]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 PAL Robotics SL.
```
Comment (olivier-stasse) [inline position_velocity_effort_gain_controller/test/test_position_velocity_effort_gain_controller.hpp]: Please add:
`// Copyright 2022 LAAS-CNRS`

**Review by olivier-stasse — CHANGES_REQUESTED** [review]
Thanks a lot Maxime.
Could make the requested changes ?
We are almost there.

File: gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp (hunk 7 above)
Comment (Maxime-Fansi-laas) [reply gazebo_ros2_control_bolt/src/gazebo_bolt_system.cpp]: I have removed the comment but kept the control law. Is it okay ?

File: position_velocity_effort_gain_controller/CMakeLists.txt (hunk 13 above)
Comment (Maxime-Fansi-laas) [reply position_velocity_effort_gain_controller/CMakeLists.txt:96]: I have added this option before "if" instruction.

File: gazebo_ros2_control_bolt/include/gazebo_ros2_control_bolt/gazebo_bolt_system_interface.hpp [hunk 17]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021 Open Source Robotics Foundation, Inc.
```
Comment (olivier-stasse) [inline gazebo_ros2_control_bolt/include/gazebo_ros2_control_bolt/gazebo_bolt_system_interface.hpp:1]: Please do not remove this line.
We get inspiration from OSRF.
They should be kept as the original author.

**Review by olivier-stasse — APPROVED** [review]

**Comment by olivier-stasse** [comment]
Thanks @Maxime-Fansi-laas  !

**Pull request merged**